File: greedy/04_Task_Scheduler/04_Task_Scheduler.js
function leastInterval(tasks, n) {
  if (n === 0) {
    return tasks.length;
  }

  const freq = new Array(26).fill(0);
  for (const task of tasks) {
    freq[task.charCodeAt(0) - "A".charCodeAt(0)]++;
  }

  let maxFreq = 0;
  let maxFreqCount = 0;
  for (const f of freq) {
    if (f > maxFreq) {
      maxFreq = f;
      maxFreqCount = 1;
    } else if (f === maxFreq) {
      maxFreqCount++;
    }
  }

  return Math.max(tasks.length, (maxFreq - 1) * (n + 1) + maxFreqCount);
}

const testCases = [
  { tasks: ["A", "A", "A", "B", "B", "B"], n: 2, expect: 8 },
  { tasks: ["A", "A", "A", "B", "B", "B"], n: 0, expect: 6 },
  {
    tasks: ["A", "A", "A", "A", "A", "A", "B", "C", "D", "E", "F", "G"],
    n: 2,
    expect: 16,
  },
  { tasks: ["A", "A", "A", "B", "B", "B", "C", "C"], n: 2, expect: 8 },
  { tasks: ["A", "A", "B", "B"], n: 3, expect: 6 },
  { tasks: ["C", "E", "C", "E", "E", "C"], n: 2, expect: 8 },
  { tasks: ["C", "E", "C", "E", "E", "C"], n: 3, expect: 10 },
  { tasks: ["A", "A", "A", "B", "B", "B", "C", "D", "E"], n: 2, expect: 9 },
  { tasks: ["A", "A", "A", "B", "B", "B", "C", "D", "E"], n: 3, expect: 11 },
  { tasks: ["A", "A", "A", "A", "B", "B", "C", "C", "D"], n: 2, expect: 11 },
  { tasks: ["A", "A", "A", "B", "B"], n: 2, expect: 7 },
];

for (let i = 0; i < testCases.length; i++) {
  const tc = testCases[i];
  console.log(
    `Test Case ${i + 1}: tasks=${JSON.stringify(tc.tasks)}, n=${tc.n}`
  );
  const res = leastInterval(tc.tasks, tc.n);
  console.log(`Got: ${res}`);
  console.log(`Expected: ${tc.expect}`);
  if (res !== tc.expect) {
    console.log("FAIL");
  } else {
    console.log("PASS");
  }
  console.log();
}

File: greedy/04_Task_Scheduler/go/04_Task_Scheduler.go
package main

import (
	"fmt"
	"math"
)

func leastInterval(tasks []byte, n int) int {
	if n == 0 {
		return len(tasks)
	}

	freq := make([]int, 26)
	for _, task := range tasks {
		freq[task-'A']++
	}

	maxFreq := 0
	maxFreqCount := 0
	for _, f := range freq {
		if f > maxFreq {
			maxFreq = f
			maxFreqCount = 1
		} else if f == maxFreq {
			maxFreqCount++
		}
	}

	return int(math.Max(float64(len(tasks)), float64((maxFreq-1)*(n+1)+maxFreqCount)))
}

func main() {
	testCases := []struct {
		tasks  []byte
		n      int
		expect int
	}{
		{[]byte{'A', 'A', 'A', 'B', 'B', 'B'}, 2, 8},
		{[]byte{'A', 'A', 'A', 'B', 'B', 'B'}, 0, 6},
		{[]byte{'A', 'A', 'A', 'A', 'A', 'A', 'B', 'C', 'D', 'E', 'F', 'G'}, 2, 16},
		{[]byte{'A', 'A', 'A', 'B', 'B', 'B', 'C', 'C'}, 2, 8},
		{[]byte{'A', 'A', 'B', 'B'}, 3, 6},
		{[]byte{'C', 'E', 'C', 'E', 'E', 'C'}, 2, 8},
		{[]byte{'C', 'E', 'C', 'E', 'E', 'C'}, 3, 10}, // Example from the prompt
		{[]byte{'A', 'A', 'A', 'B', 'B', 'B', 'C', 'D', 'E'}, 2, 9},
		{[]byte{'A', 'A', 'A', 'B', 'B', 'B', 'C', 'D', 'E'}, 3, 11},
		{[]byte{'A', 'A', 'A', 'A', 'B', 'B', 'C', 'C', 'D'}, 2, 11},
	}

	for i, tc := range testCases {
		fmt.Printf("Test Case %d: tasks=%s, n=%d\n", i+1, string(tc.tasks), tc.n)
		res := leastInterval(tc.tasks, tc.n)
		fmt.Printf("Got: %d\n", res)
		fmt.Printf("Expected: %d\n", tc.expect)
		if res != tc.expect {
			fmt.Println("FAIL")
		} else {
			fmt.Println("PASS")
		}
		fmt.Println()
	}
}
